### PR TITLE
Adding ec2 types

### PR DESCRIPTION
**Pull request by nhibberd**
@markhibberd Exposing some more EC2 types at this level to be re-used upstream
- `mismi-ec2` now has a min amazonka bounds of 1.4.0
- Exposing instance types as a mismi data type
- moved everything i could back into `mismi-ec2-core`

Will follow this up with a similar PR for `IAM` and `Autoscaling`

### CODE REVIEW

File: mismi-ec2-core/src/Mismi/EC2/Core/Data.hs [hunk 1]
```diff
@@ -1,4 +1,278 @@
...
+
+data EC2Market =
+    OnDemand
+  | Spot Text MismiSpotInstanceType
+  deriving (Eq, Show)
+
+data EC2Tag =
+  EC2Tag {
+      tagKey :: Text
+    , tagVale :: Text
```
Comment (olorin) [inline mismi-ec2-core/src/Mismi/EC2/Core/Data.hs]: Bangs?

File: mismi-ec2-core/src/Mismi/EC2/Core/Data.hs [hunk 2]
```diff
@@ -1,4 +1,278 @@
...
+    } deriving (Eq, Show, Ord)
+
+newtype AvailabilityZone =
+  AvailabilityZone {
+      availabilityZone :: Text
+    } deriving (Eq, Show, Ord)
+
+data EC2Market =
+    OnDemand
+  | Spot Text MismiSpotInstanceType
```
Comment (olorin) [inline mismi-ec2-core/src/Mismi/EC2/Core/Data.hs]: Bangs?

**Comment by olorin** [comment]
LGTM, :+1: on :green_apple:

File: mismi-ec2/src/Mismi/EC2/Data.hs [hunk 3]
```diff
@@ -1,65 +1,259 @@
...
+    T2_Large ->
+      A.T2_Large
+    T2_Medium ->
+      A.T2_Medium
+    T2_Micro ->
+      A.T2_Micro
+    T2_Nano ->
+      A.T2_Nano
+    T2_Small ->
+      A.T2_Small
```
Comment (markhibberd) [inline mismi-ec2/src/Mismi/EC2/Data.hs]: Is this going to have to be bidirectional? As in do you want a mapping the other way. Would be good to add as a test regardless, but feel like you will hit a wall without that mapping as well.

Comment (markhibberd) [inline mismi-ec2/src/Mismi/EC2/Data.hs]: Derp, ignore me, glossed over diff below.

File: mismi-ec2/test/test.hs [hunk 4]
```diff
@@ -1,9 +1,6 @@
 import           Disorder.Core.Main
 
-import           Test.Mismi.EC2.Data
-
 main :: IO ()
 main =
   disorderMain [
-      Test.Mismi.EC2.Data.tests
```
Comment (markhibberd) [inline mismi-ec2/test/test.hs]: Tests for conversions to and fro?

File: mismi-ec2/test/Test/IO/Mismi/EC2/Arbitrary.hs [hunk 5]
```diff
@@ -9,10 +9,6 @@ import           P
 
 import           Test.QuickCheck
 
-instance Arbitrary VirtualizationType where
-  arbitrary =
-    elements [HVM, Paravirtual]
-
 instance Arbitrary SecurityGroupName where
```
Comment (markhibberd) [inline mismi-ec2/test/Test/IO/Mismi/EC2/Arbitrary.hs]: Shouldn't this go to core arbitrary as well (and possibly others). Plus can this pull in the core instances so you dont have to depend on both everytime.

**Comment by markhibberd** [comment]
Few comments inline. 👍 after they are worked through.

File: mismi-ec2/test/test.hs (hunk 4 above)
Comment (nhibberd) [inline mismi-ec2/test/test.hs]: In core now

Comment (nhibberd) [inline mismi-ec2/test/test.hs]: Will add instance-type tests

**Pull request merged**